### PR TITLE
Buffon's Needle

### PR DESCRIPTION
**Pull request by VPraharsha03**

### CODE REVIEW

File: book_programs/chapter_2/buffons_needle.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
+# Usage: python3 name_of_program.py <n>
```
Comment (FireLemons) [inline book_programs/chapter_2/buffons_needle.py]: change `name_of_program` to `buffons_needle`

**Comment by FireLemons** [comment]
It'll take me a little while to review this. I'll probably finish some time tomorrow

File: book_programs/chapter_2/buffons_needle.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+# Arg is negative
+if 'n' in globals() and n < 0:
+    bad_input = True
+    print("ERROR: argument cannot be negative")
+
+# Print usage help when arguments are bad
+if bad_input:
+    print("\nUsage: python3 buffons_needle.py <n>")
+    print("  n:   The Number of Needles used")
```
Comment (FireLemons) [inline book_programs/chapter_2/buffons_needle.py:41]: also print the example usage

File: book_programs/chapter_2/buffons_needle.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        self.list_of_needle_objects.append(needle_object)
+        x_coordinates = [needle_object.end_points[0]
+                         [0], needle_object.end_points[1][0]]
+        y_coordinates = [needle_object.end_points[0]
+                         [1], needle_object.end_points[1][1]]
+
+        for board in range(self.boards):
+            if needle_object.intersects_with_y(self.floor[board]):
+                self.number_of_intersections += 1
+                self.buffon.plot(x_coordinates, y_coordinates,
```
Comment (FireLemons) [inline book_programs/chapter_2/buffons_needle.py:97]: In the mathematica file not every needle is plotted. Only every 100th needle is plotted so the simulation can support n=10000

File: book_programs/chapter_2/buffons_needle.py (hunk 1 above)
Comment (FireLemons) [inline book_programs/chapter_2/buffons_needle.py]: The mathematica version has a print boolean parameter which determines whether the results are graphed. The estimate for pi is always printed in the console. So you could run the experiment and get the estimate for pi without seeing the graph.

File: book_programs/chapter_2/buffons_needle.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        pyplot.title("The simulated estimate for Pi is ")
+
+    def plot(self):
+        self.plot_floor_boards()
+        self.plot_needles()
+        pyplot.show()
+
+
+simulation = BuffonSimulation()
+simulation.plot()
```
Comment (FireLemons) [inline book_programs/chapter_2/buffons_needle.py:134]: the estimate for pi is always printed in the console.

**Review by FireLemons — CHANGES_REQUESTED** [review]

**Comment by FireLemons** [comment]
@VPraharsha03 buffons_needle.py works great. Would you like to improve it a bit further? I can merge and create new issues from this if you'd like to work on other things.

File: book_programs/chapter_2/buffons_needle.py (hunk 1 above)
Comment (VPraharsha03) [reply book_programs/chapter_2/buffons_needle.py]: I was thinking about that. I'll look into it. I'm AFK for a while.

**Comment by VPraharsha03** [comment]
> It'll take me a little while to review this. I'll probably finish some time tomorrow

It will take me time and will be late, as I'm busy with things.

**Comment by FireLemons** [comment]
@VPraharsha03 Oh I've finished reviewing. Would you like to merge these changes and move on to other things?

**Comment by VPraharsha03** [comment]
@FireLemons OK. If everything seems fine to you, you can go ahead. I will make the rest of the changes at a later time.

**Comment by VPraharsha03** [comment]
> @VPraharsha03 Oh I've finished reviewing. Would you like to merge these changes and move on to other things?

@FireLemons Yes please.

**Comment by VPraharsha03** [comment]
@FireLemons Will it be marked #hacktober-accepted after I complete the pending improvements ? (because the issue is still open)

**Pull request merged**